test/filesystem.cc

```diff
@@ -135,6 +137,70 @@ TEST(Filesystem, extension) {
 TEST(Filesystem, file_size) {
 }
 
+TEST(Filesystem, concurrent_folder_create_delete) {
+
+  const char* create_subdir = "test/a/b/c/d/e/f/g/h/i/j";
+  const char* base_subdir = "test/a";
+
+  auto create_folder = [&]() {
+    bool success = filesystem::create_directories(create_subdir);
+    EXPECT_TRUE(success);
+    EXPECT_TRUE(filesystem::exists(create_subdir));
+  };
+
+  auto remove_folder = [&]() {
+    std::uintmax_t delete_count = filesystem::remove_all(base_subdir);
+    // I've found that two+ threads will claim that they've deleted the same
+    // file object. This results in double+ counting and prevents me from
+    // asserting anything related to the delete_count. All we can be sure
+    // is the following:
+    EXPECT_TRUE(!filesystem::exists(base_subdir));
+  };
+
+  size_t num_threads = std::thread::hardware_concurrency();
+
+  const int num_iters = 50;
+  for (int j = 0; j < num_iters; j++) {
+    // concurrent folder creation
+    {
+      std::vector<std::unique_ptr<std::thread>> threads(num_threads);
+      for (size_t i = 0; i < num_threads; i++)
+        threads[i].reset(new std::thread(create_folder));
```

Comment:
There's a risk that no concurrency at all happens if each thread finishes before the next is created since there is no synchronization. I guess the filesystem operations likely prevent this from happening.

Reply:
Right, I'll implement a "starting gun" event.  Thanks.

Reply:
I think since the thread is making a syscall (to do file IO), its highly likely that other processes (and in this case threads) get scheduled in during that time.

Reply:
Mandeep is probably right - the file i/o slows things down enough that concurrency is almost guaranteed.  That said, Anders point is still valid - and there's no harm in sophisticating the test a bit to maximize concurrency.  Hence my addition of the "start_race_event" condition variable.